[Concurrency] Correct isolation handling with default act and task exec

ktoso opened the pull request:
We did not properly account for default actor's isolation in our task executor forwarding and... this is quite unfortunate.

~~TODO: rebase it on top of https://github.com/swiftlang/swift/pull/82913 which does the Task.immediate changes~~ (done)

rdar://155596073
Also rdar://144382903

al45tair commented on stdlib/public/Concurrency/Executor.swift:
```diff
@@ -461,6 +461,14 @@ public protocol TaskExecutor: Executor {
+  // Must implement this requirement in order
+  @_nonoverride
+  @available(StdlibDeploymentTarget 6.2, *)
+  func enqueue(_ job: consuming ExecutorJob, isolatedTo serialExecutor: UnownedSerialExecutor/*some SerialExecutor*/)
```
Should there be an implementation of this in `DispatchExecutor.swift`? It's not obvious to me what the `isolatedTo:` argument actually does here, or how an executor implementation would use it. Maybe some doc comments might help?

ktoso commented:
I'm reworking this to better handle the task executor in default actor enqueue, we may not need the extra requirement.

ktoso committed:
[Concurrency] Honor task executor preference when default actor is involved

When a default actor and a custom task executor preference coexisted, the
runtime dropped the preference in three places:

- swift_task_enqueueImpl: only the generic-serial-executor branch honored
  a task executor preference, so jobs targeting a default actor lost it.
- DefaultActorImpl::enqueue: the actor's enqueue loop only loaded the task
  executor preference when it thought it would need to schedule, so an
  already-scheduled actor would never hop to the requested task executor.
- swift_job_runImpl: always force-unlocked the actor on switch out, which
  conflicted with a task executor that wanted to keep draining the actor.

This patch routes default-actor + task-executor combos through the actor's
own enqueue (which now always reads the preference and re-schedules a
processing job when one is set), and leaves the actor lock held when a
task executor is in play so the next pending job lands on the right
executor.

swift_task_immediateImpl gains a default-actor branch: an immediate task
targeting a default actor must not run synchronously on the caller; it is
enqueued on the actor so the actor's drain (with the preference applied)
runs it on the right thread.

Renames forSynchronousStart -> forImmediateTask and isForSynchronousStart
-> isImmediateTask for clarity, and adds Task_IsImmediateTask job-flag
plumbing so the runtime can distinguish immediate-task starts elsewhere.

rdar://155596073

ktoso commented:
I'm slowly reviving/rebasing this, not ready for review at all; just getting back to a clean state here, it was too long ago

ktoso commented:
I'll just close this fornow
